### PR TITLE
CAMEL-14326 Update to introduce the camel-qute component

### PR DESCRIPTION
**Pull request by zhfeng**
https://issues.apache.org/jira/browse/CAMEL-14326

The following work could be to support the ValueResolver, Inject beans, customs tags and etc.

### CODE REVIEW

File: core/camel-componentdsl/src/generated/resources/metadata.json [hunk 1]
```diff
@@ -5263,6 +5263,26 @@
...
+    "description": "Transforms the message using a Quarkus Qute template.",
+    "label": "transformation",
+    "deprecated": false,
+    "deprecationNote": "transformation",
+    "async": false,
+    "consumerOnly": false,
+    "producerOnly": true,
+    "lenientProperties": false,
+    "javaType": "org.apache.camel.component.qute.QuteComponent",
+    "firstVersion": "3.1.0",
```
Comment (davsclaus) [inline core/camel-componentdsl/src/generated/resources/metadata.json]: Should be 3.2.0

File: components/camel-qute/src/generated/resources/qute.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "other": {
+    "kind": "other",
+    "name": "qute",
+    "title": "Qute",
+    "description": "Camel Qute support",
+    "deprecated": false,
+    "groupId": "org.apache.camel",
+    "artifactId": "camel-qute",
+    "version": "3.1.0-SNAPSHOT"
```
Comment (oscerd) [inline components/camel-qute/src/generated/resources/qute.json]: Here too, it should be 3.2.0-SNAPSHOT

File: components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    public QuteEndpoint() {
+    }
+
+    public QuteEndpoint(String uri, QuteComponent component, String resourceUri) {
+        super(uri, component, resourceUri);
+    }
+
+    @Override
+    public ExchangePattern getExchangePattern() {
+        return ExchangePattern.InOut;
```
Comment (oscerd) [inline components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java]: Is there any reason for this? You can already override the exchangePattern as option if you want.

File: components/camel-qute/src/test/java/org/apache/camel/component/qute/QuteTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        assertEquals("Dear Christian. You ordered item 7 on Monday.", exchange.getMessage().getBody());
+        assertEquals("Christian", exchange.getMessage().getHeader("name"));
+    }
+
+    @Override
+    protected RoutesBuilder createRouteBuilder() throws Exception {
+        return new RouteBuilder() {
+            public void configure() {
+                // START SNIPPET: example
```
Comment (oscerd) [inline components/camel-qute/src/test/java/org/apache/camel/component/qute/QuteTest.java]: I think the snippet comments could be removed.

**Comment by oscerd** [comment]
Also please fix the codestyle by running the build with -Psourcecheck. I do believe you need to completely rebuild camel with 3.2.0-SNAPSHOT first

File: components/camel-qute/src/generated/java/org/apache/camel/component/qute/QuteComponentConfigurer.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/* Generated by camel build tools - do NOT edit this file! */
```
Comment (dmvolod) [inline components/camel-qute/src/generated/java/org/apache/camel/component/qute/QuteComponentConfigurer.java]: Look like auto-generated files should include ASF header.

Comment (oscerd) [reply components/camel-qute/src/generated/java/org/apache/camel/component/qute/QuteComponentConfigurer.java]: We excluded them from the checkstyle.

File: components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java (hunk 3 above)
Comment (zhfeng) [reply components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java]: No, it should be removed and I just borrow this from the camel-velocity.

File: components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteComponent.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.annotations.Component;
+import org.apache.camel.support.DefaultComponent;
+import org.apache.camel.support.ResourceHelper;
+
+@Component("qute")
+public class QuteComponent extends DefaultComponent {
+
+    @Metadata(label = "advanced")
+    private Engine quarkusEngine;
```
Comment (davsclaus) [inline components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteComponent.java]: This should be named quteEngine

File: components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                    if (path.equals(getResourceUri())) {
+                        in = getResourceAsInputStream();
+                    } else {
+                        in = locatePath(path).openStream();
+                    }
+
+                    Reader reader = getEncoding() != null ? new InputStreamReader(in, getEncoding()) : new InputStreamReader(in);
+                    return reader;
+                } catch (Exception e) {
+                    log.warn("can not load template " + path + " due to " + e);
```
Comment (davsclaus) [inline components/camel-qute/src/main/java/org/apache/camel/component/qute/QuteEndpoint.java]: Cannot

File: components/camel-qute/src/main/docs/qute-component.adoc [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+| *contentCache* (producer) | Sets whether to use resource content cache or not | false | boolean
+| *encoding* (producer) | Character encoding of the resource content. |  | String
+| *lazyStartProducer* (producer) | Whether the producer should be started lazy (on the first message). By starting lazy you can use this to allow CamelContext and routes to startup in situations where a producer may otherwise fail during starting and cause the route to fail being started. By deferring this startup to be lazy then the startup failure can be handled during routing messages via Camel's routing error handlers. Beware that when the first message is processed then creating and starting the producer may take a little time and prolong the total processing time of the processing. | false | boolean
+| *basicPropertyBinding* (advanced) | Whether the endpoint should use basic property binding (Camel 2.x) or the newer property binding with additional capabilities | false | boolean
+| *synchronous* (advanced) | Sets whether synchronous processing should be strictly used, or Camel is allowed to use asynchronous processing (if supported). | false | boolean
+|===
+// endpoint options: END
```
Comment (davsclaus) [inline components/camel-qute/src/main/docs/qute-component.adoc]: It would be good to add a new section with one or two examples to the docs

Comment (zhfeng) [reply components/camel-qute/src/main/docs/qute-component.adoc]: yeah, I will add the examples.

**Comment by davsclaus** [comment]
@zhfeng let us know when you have the last bits done so we can get this new component merged

**Comment by zhfeng** [comment]
Thanks @davsclaus - I just added the example section and re-generate the files.

**Comment by davsclaus** [comment]
@oscerd @lburgazzoli oh I wonder if we merge this to camel repo, and have a dependency on quarkus-qute, and then we have a camel-quarkus extension for this component, and then we wont end up with a circular quakus dependency problem. Or a N-1 problem. eg camel-qute uses quarkus 1.2.1, and then quarkus 1.3.0 is released, and we want to release camel-quarkus with that version, and before we could just do it, now we may need to release camel first to pickup quarkus 1.3.0, and then thereafter only camel-quarkus.

So maybe camel-qute should be moved to camel-quarkus and be only there?

**Comment by oscerd** [comment]
I think it makes sense to have it in camel-quarkus. You're right

**Comment by zhfeng** [comment]
yeah, there is also an issue with [Qute support](https://github.com/apache/camel-quarkus/issues/825) in camel-quarkus. So I think it should be OK to move there.

**Comment by zhfeng** [comment]
I close this PR and move the camel-qute component to the camel-quarkus.

**Pull request closed without merging**